File: ccg/src/db/validator.pl
#
# Form validation functions
#

use POSIX;

if ( $ENV{'CCG_PERLLIBDIR'} ne '' )
{ $CCG_PERLLIBDIR = $ENV{'CCG_PERLLIBDIR'}; }
else
{ $CCG_PERLLIBDIR = '/projects/src'; }

require "$CCG_PERLLIBDIR/db/validator.pl";

#
##################################
# Field validating functions
##################################
#
sub isBlank
{
   my ($val) = @_;
   my $char;

   # This is similar to the empty() check in PHP
   if ( $val eq '' || $val eq '0' || &isNULL($val) ) { return 1; }

   foreach (split //, $val)
   {
      $char = $_;

      if ( $char =~ m/[^ \t\n\r]/ ) { return 0; }
   }

   return 1;
}

sub ValidDate
{
   my ($d_val) = @_;

   if ( &isBlank($d_val) ) { return 0; }

   $d_val = &trim($d_val);

   if ( $d_val !~ m/^[0-9]{4}\-[0-9]{1,2}\-[0-9]{1,2}$/ ) { return 0; }

   my ($yr, $mo, $dy) = split(/\-/, $d_val);

   # Allow default date
   if ( $yr eq '0000' and $mo eq '00' and $dy eq '00' ) { return 1; }

   # Check year
   if ( ! &ValidinRange($yr,'1900','9999','int') ) { return 0; }

   # Check month
   if ( ! &ValidinRange($mo,'1','12','int') ) { return 0; }

   # Check day

   # First entry is 0 so that the index matches the month
   @lastdom = ( 0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 );

   my $leap = &leapyear($yr);
   if ( $leap ) { $lastdom[2] = 29; }

   $lastday = $lastdom[$mo];

   if ( ! &ValidinRange($dy,'1',$lastday,'int') ) { return 0; }

   return 1;   
}

sub ValidDatetime
{
   my ($dt_val) = @_;

   if ( &isBlank($dt_val) ) { return 0; }

   $dt_val = &trim($dt_val);

   my ( $date, $time ) = split (/\s/, $dt_val);

   if ( &ValidDate($date) && &ValidTime($time) )
   { return 1; }
   else
   { return 0; }
}

sub ValidFloat
{
   my ($flt_val) = @_;

   $flt_val = &trim($flt_val);

   # The regular expression reads
   # Match the string that begins with 0 or 1 -'s
   # Then 0 or more characters between 0 and 9 followed by
   #    0 or 1 periods followed by 1 or more characters between 0 and 9
   # OR
   # 1 or more characters between 0 and 0 followed by
   #    0 or 1 periods followed by 0 or more characters between 0 and 9
   if ( $flt_val =~ m/^-?([0-9]*\.?[0-9]+|[0-9]+\.?[0-9]*)$/ ) { return 1; }

   return 0;
}

sub ValidinRange
{
   my ($value, $min, $max, $type ) = @_;

   $value = &trim($value);
   $min = &trim($min);
   $max = &trim($max);
   $type = &trim($type);

   if ( &isBlank($type) ) { $type = "other"; }

   if ( $type eq 'date' )
   {
      if ( ! &ValidDate($value) ) { return 0; }

      my ($yr, $mo, $dy) = split(/\-/, $value);

      my $dec = &date2dec($yr, $mo, $dy, 12, 0);

      if ( ! &isNULL($min) )
      {
         if ( ! &ValidDate($min) ) { return 0; }

         my ($minyr, $minmo, $mindy) = split(/\-/, $min);
         my $mindec = &date2dec($minyr, $minmo, $mindy, 12, 0);
         if ( $dec < $mindec ) { return 0; }
      }

      if ( ! &isNULL($max) )
      {
         if ( ! &ValidDate($max) ) { return 0; }

         my ($maxyr, $maxmo, $maxdy) = split(/\-/, $max);
         my $maxdec = &date2dec($maxyr, $maxmo, $maxdy, 12, 0);
         if ( $dec > $maxdec ) { return 0; }
      }
      return 1;
   }
   elsif ( $type eq 'time' )
   {
      if ( ! &ValidTime($value) ) { return 0; }

      my ($hr, $mn, $sc) = split(/:/, $value);

      my $dec = &date2dec(2000, 1, 1, $hr, $mn, $sc);

      if ( ! &isNULL($min) )
      {
         if ( ! &ValidTime($min) ) { return 0; }

         my ($minhr, $minmn, $minsc) = split(/\-/, $min);
         my $mindec = &date2dec(2000, 1, 1, $minhr, $minmn, $minsc);
         if ( $dec < $mindec ) { return 0; }
      }

      if ( ! &isNULL($max) )
      {
         if ( ! &ValidDate($max) ) { return 0; }

         my ($maxhr, $maxmn, $maxsc) = split(/\-/, $max);
         my $maxdec = &date2dec(2000, 1, 1, $maxhr, $maxmn, $maxsc);
         if ( $dec > $maxdec ) { return 0; }
      }
      return 1;
   }
   elsif ( $type eq 'int' )
   {
      if ( ! &ValidInt($value) ) { return 0; }

      if ( ! &isNULL($min) )
      {
         if ( ! &ValidInt($min) ) { return 0; }
         if ( $value < $min ) { return 0; }
      }

      if ( ! &isNULL($max) )
      {
         if ( ! &ValidInt($max) ) { return 0; }
         if ( $value > $max ) { return 0; }
      }

      return 1;
   }
   elsif ( $type eq 'float' )
   {
      if ( ! &ValidFloat($value) ) { return 0; }

      if ( ! &isNULL($min) )
      {
         if ( ! &ValidFloat($min) ) { return 0; }
         if ( $value < $min ) { return 0; }
      }

      if ( ! &isNULL($max) )
      {
         if ( ! &ValidFloat($max) ) { return 0; }
         if ( $value > $max ) { return 0; }
      }
      return 1;

   }
   elsif ( $type eq 'char' )
   {
      if ( ! &isNULL($min) )
      { if ( $value lt $min ) { return 0; } }
      if ( ! &isNULL($max) )
      { if ( $value gt $max ) { return 0; } }
      return 1;
   }
   else
   {
      if ( ! &isNULL($min) )
      { if ( $value < $min ) { return 0; } }
      if ( ! &isNULL($max) )
      { if ( $value > $max ) { return 0; } }
      return 1;
   }

   return 0;
}

sub ValidInt
{
   my ($int_val) = @_;

   $int_val = &trim($int_val);

   if ( $int_val =~ m/^-?[0-9]+$/ ) { return 1; }

   return 0;
}

sub ValidLength
{
   my ($value, $min, $max) = @_;

   $value = &trim($value);
   $min = &trim($min);
   $max = &trim($max);

   if ( ! &ValidInt($min) ) { return 0; }
   if ( ! &ValidInt($max) ) { return 0; }

   if ( length($value) < $min ) { return 0; }
   if ( length($value) > $max ) { return 0; }

   return 1;
}

sub ValidTime
{
   my ($t_val) = @_;

   if ( &isBlank($t_val) ) { return 0; }

   $t_val = &trim($t_val);

   # If only the hour and minute were specified, add the seconds
   if ( $t_val =~ m/^[0-9]{1,2}:[0-9]{2}$/ )
   { $t_val = $t_val.":00"; }

   if ( $t_val !~ /^[0-9]{1,2}:[0-9]{2}:[0-9]{2}$/ )
   { return 0; }

   my ($hr, $mn, $sc) = split(/:/, $t_val);

   if ( ! &ValidinRange($hr,"0","23","int") ) { return 0; }
   if ( ! &ValidinRange($mn,"0","59","int") ) { return 0; }
   if ( ! &ValidinRange($sc,"0","59","int") ) { return 0; }

   ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst) = localtime (time);

   ($chksc, $chkmn, $chkhr, $junk) = localtime(mktime ($sc, $mn, $hr, $mday, $mon, $year, $wday, $yday, $isdst));

   $t_chk1 = sprintf("%02d:%02d:%02d", $chkhr, $chkmn, $chksc);
   $t_chk2 = sprintf("%1d:%02d:%02d", $chkhr, $chkmn, $chksc);

   if ( $t_val eq $t_chk1 || $t_val eq $t_chk2 ) { return 1; }
   return 0;
}

#
##################################
# Single character validating functions
##################################
#
sub isAlphaNum
{
   my ($val) = @_;

   if ( &isLetter($val) || &isDigit($val) ) { return 1; }
   return 0;
}

sub isDigit
{
   my ($val) = @_;

   if ( length($val) > 1 ) { return 0; }
   if ( $val =~ m/^[0-9]$/ ) { return 1; }
   return 0;
}

sub isLetter
{
   my ($val) = @_;

   if ( &isLowerLetter($val) || &isUpperLetter($val) ) { return 1; }
   return 0;
}

sub isNULL
{
   my ($val) = @_;

   if ( $val eq "NULL" ) { return 1; }
   return 0;
}

sub isLowerLetter
{
   my ($val) = @_;

   if ( length($val) > 1 ) { return 0; }

   if ( $val =~ /^[a-z]$/ ) { return 1; }

   return 0;
}

sub isUpperLetter
{
   my ($val) = @_;

   if ( length($val) > 1 ) { return 0; }

   if ( $val =~ /^[A-Z]$/ ) { return 1; }

   return 0;
}

#
#####################################
# Miscellaneous functions
#####################################
#
sub inString
{
   my ($str, $cstr) = @_;

   $pos = index($str, $cstr);
   if ( $pos >= 0 ) { return 1; }
   return 0;
}

1;

File: web_applications/inc/validator.php
<?PHP
#
# Form validation functions
#

#
########################################################################################
# Field validating functions
########################################################################################
#
function isBlank($val)
{
   # empty checks for "", 0, "0", NULL, FALSE, array(),
   #    and var $var [declared variable without a value]
   if ( strlen($val) == 0 ) { return TRUE; }

   $val_arr = str_split($val);

   for ( $i=0; $i<count($val_arr); $i++ )
   {
      if ( preg_match("/[^ \t\n\r]/", $val_arr[$i] ) ) { return FALSE; }
   }

   return TRUE;
}
function ValidDate($d_val)
{
   if ( isBlank($d_val) ) { return FALSE; }

   if ( ! is_string($d_val) ) { return FALSE; }

   if ( ! preg_match("/^[0-9]{4}\-[0-9]{1,2}\-[0-9]{1,2}$/", $d_val) )
   { return FALSE; }

   list($yr, $mo, $dy) = explode("-", $d_val);

   $mindate = MinDate();
   $maxdate = MaxDate();

   list($mindateyr, $mindatemo, $mindatedy) = explode("-", $mindate);
   list($maxdateyr, $maxdatemo, $maxdatedy) = explode("-", $maxdate);

   if ( ! ValidinRange($yr,$mindateyr,$maxdateyr,"int") ) { return FALSE; }
   if ( ! ValidinRange($mo,$mindatemo,$maxdatemo,"int") ) { return FALSE; }
   if ( ! ValidinRange($dy,$mindatedy,$maxdatedy,"int") ) { return FALSE; }

   if ( ! checkdate($mo, $dy, $yr) ) { return FALSE; }

   return TRUE;
}

function ValidDatetime($dt_val)
{
   if ( isBlank($dt_val) ) { return FALSE; }

   if ( ! is_string($dt_val) ) { return FALSE; }

   if ( ! preg_match("/^[0-9]{4}\-[0-9]{1,2}\-[0-9]{1,2} [0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}$/", $dt_val) )
   { return FALSE; }

   list($date, $time) = preg_split('/\s/', $dt_val, 2);

   if ( ValidDate($date) && ValidTime($time) )
   { return TRUE; }
   else
   { return FALSE; }
}

function ValidFloat($flt_val)
{
   if ( ! is_numeric($flt_val) ) { return FALSE; }

   if ( preg_match("/^[+-]?[0-9]*\.?[0-9]+$/", $flt_val) ) { return TRUE; }
   return FALSE;
}

function ValidinRange($value, $min, $max, $type="other")
{
   switch ( $type )
   {
      case "date":
         if ( ! ValidDate($value) ) { return FALSE; }

         list($yr, $mo, $dy) = explode("-", $value);

         $dec = Date2Dec($yr, $mo, $dy, 12, 0);

         if ( ! isNULL($min) )
         {
            if ( ! ValidDate($min) ) { return FALSE; }

            list($minyr, $minmo, $mindy) = explode("-", $min);
            $mindec = Date2Dec($minyr, $minmo, $mindy, 12, 0);
            if ( $dec < $mindec ) { return false; }
         }

         if ( ! isNULL($max) )
         {
            if ( ! ValidDate($max) ) { return FALSE; }

            list($maxyr, $maxmo, $maxdy) = explode("-", $max);
            $maxdec = Date2Dec($maxyr, $maxmo, $maxdy, 12, 0);
            if ( $dec > $maxdec ) { return false; }
         }
         break;
      case "int":
         if ( ! ValidInt($value) ) { return FALSE; }

         if ( ! isNULL($min) )
         {
            if ( ! ValidInt($min) ) { return FALSE; }
            if ( intval($value) < intval($min) ) { return FALSE; }
         }

         if ( ! isNULL($max) )
         {
            if ( ! ValidInt($max) ) { return FALSE; }
            if ( intval($value) > intval($max) ) { return FALSE; }
         }
         break;
      case "float":
         if ( ! ValidFloat($value) ) { return FALSE; }

         if ( ! isNULL($min) )
         {
            if ( ! ValidFloat($min) ) { return FALSE; }
            if ( floatval($value) < floatval($min) ) { return FALSE; }
         }

         if ( ! isNULL($max) )
         {
            if ( ! ValidFloat($max) ) { return FALSE; }
            if ( floatval($value) > floatval($max) ) { return FALSE; }
         }
         break;
      default:
         if ( ! isNULL($min) )
         { if ( $value < $min ) { return FALSE; } }
         if ( ! isNULL($max) )
         { if ( $value > $max ) { return FALSE; } }
         break;
   }
   return TRUE;
}

function ValidInt($int_val)
{
   if ( ! is_numeric($int_val) ) { return FALSE; }

   if ( preg_match("/^[+-]?[0-9]+$/", $int_val ) ) { return TRUE; }
   return FALSE;
}

function ValidLength($value, $min, $max)
{
   if ( ! ValidInt($min) ) { return FALSE; }
   if ( ! ValidInt($max) ) { return FALSE; }

   if ( strlen($value) < $min ) { return FALSE; }
   if ( strlen($value) > $max ) { return FALSE; }
   return TRUE;
}

function ValidTime($t_val)
{
   if ( isBlank($t_val) ) { return FALSE; }

   if ( ! is_string($t_val) ) { return FALSE; }

   # If only the hour and minute were specified, add the seconds
   if ( preg_match("/^[0-9]{1,2}:[0-9]{2}$/", $t_val) )
   { $t_val = $t_val.":00"; }

   if ( ! preg_match("/^[0-9]{1,2}:[0-9]{2}:[0-9]{2}$/", $t_val) )
   { return FALSE; }

   list($yr, $mo, $dy) = explode("-", date("Y-m-d"));

   list($hr, $mn, $sc) = explode(":", $t_val);

   $mintime = MinTime();
   $maxtime = MaxTime();

   list($mintimehr, $mintimemn, $mintimesc) = explode(":", $mintime);
   list($maxtimehr, $maxtimemn, $maxtimesc) = explode(":", $maxtime);

   if ( ! ValidinRange($hr,$mintimehr,$maxtimehr,"int") ) { return FALSE; }
   if ( ! ValidinRange($mn,$mintimemn,$maxtimemn,"int") ) { return FALSE; }
   if ( ! ValidinRange($sc,$mintimesc,$maxtimesc,"int") ) { return FALSE; }

   # With 0 padding
   $t_chk1 = date("H:i:s", mktime($hr, $mn, $sc, $mo, $dy, $yr));
   # Without 0 padding
   $t_chk2 = date("G:i:s", mktime($hr, $mn, $sc, $mo, $dy, $yr));

   if ( $t_val === $t_chk1 || $t_val === $t_chk2 ) { return TRUE; }
   return FALSE;
}

#
########################################################################################
# Single character validating functions
########################################################################################
#
function isAlphaNum($val)
{
   if ( ! isLetter($val) && ! isDigit($val) ) { return FALSE; }
   return TRUE;
}


function isDigit($val)
{
   if ( strlen($val) > 1 ) { return FALSE; }
   if ( ! preg_match("/^[0-9]$/", $val) ) { return FALSE; }
   return TRUE;
}

function isLetter($val)
{
   if ( ! isLowerLetter($val) && ! isUpperLetter($val) )
   { return FALSE; }
   return TRUE;
}

function isNULL($val)
{
   if ( is_null($val) ) { return TRUE; }
   return FALSE;
}

function isLowerLetter($val)
{
   if ( strlen($val) > 1 ) { return FALSE; }
   if ( ! preg_match("/^[a-z]$/", $val) ) { return FALSE; }
   return TRUE;
}

function isUpperLetter($val)
{
   if ( strlen($val) > 1 ) { return FALSE; }
   if ( ! preg_match("/^[A-Z]$/", $val) ) { return FALSE; }
   return TRUE;
}

#
########################################################################################
# Minimum & maximum functions
########################################################################################
#
function MaxDate($value='')
{
   if ( preg_match('/^[0-9]{4}$/', $value ) )
   {
      $value = $value.'-12-31';
   }
   elseif ( preg_match('/^[0-9]{4}\-[0-9]{1,2}$/', $value ) )
   {
      // Days in month
      $dim_noleap = array ('0','31','28','31','30','31','30','31','31','30','31','30','31');
      $dim_leap = array ('0','31','29','31','30','31','30','31','31','30','31','30','31');

      $fields = explode('-', $value);
      $yr = $fields[0];
      $mo = $fields[1];

      if (($yr%4==0 && $yr%100 != 0) || ($yr%400 == 0))
      { $dy = $dim_leap[$mo]; }
      else
      { $dy = $dim_noleap[$mo]; }

      // Add day
      $value = $value.'-'.$dy;
   }
   else
   {
      $value = '9999-12-31';
   }

   return $value;
}

function MinDate($value='')
{
   if ( preg_match('/^[0-9]{4}$/', $value ) )
   {
      $value = $value.'-1-1';
   }
   elseif ( preg_match('/^[0-9]{4}\-[0-9]{1,2}$/', $value ) )
   {
      $value = $value.'-1';
   }
   else
   {
      $value = '1900-1-1';
   }

   return $value;
}

function MaxTime($value='')
{
   if ( preg_match('/^[0-9]{1,2}$/', $value ) )
   {
      $value = $value.':59:59';
   }
   elseif ( preg_match('/^[0-9]{1,2}:[0-9]{2}$/', $value ) )
   {
      $value = $value.':59';
   }
   else
   {
      $value = '23:59:59';
   }

   return $value;
}

function MinTime($value='')
{
   if ( preg_match('/^[0-9]{1,2}$/', $value ) )
   {
      $value = $value.':00:00';
   }
   elseif ( preg_match('/^[0-9]{1,2}:[0-9]{2}$/', $value ) )
   {
      $value = $value.':00';
   }
   else
   {
      $value = '00:00:00';
   }

   return $value;
}

function MaxLatitude()
{
   $value = floatval('90.0');
   return $value;
}

function MinLatitude()
{
   $value = floatval('-90.0');
   return $value;
}

function MaxLongitude()
{
   $value = floatval('180.0');
   return $value;
}

function MinLongitude()
{
   $value = floatval('-180.0');
   return $value;
}

function MaxHeight()
{
   $value = floatval('99999.99999');
   return $value;
}

function MinHeight()
{
   $value = floatval('-99999.99999');
   return $value;
}

#
########################################################################################
# Miscellaneous functions
########################################################################################
#
function inString($str, $cstr)
{
   $pos = strpos($str, $cstr);
   if ( $pos === false ) { return FALSE; }
   return TRUE;
}

function in_clnarray($element, $array )
{

   if ( ! is_array($array) ) { return(FALSE); }

   for ( $i=0; $i<count($array); $i++ )
   {
      if ( $array[$i] === $element ) { return(TRUE); }
   }

   return(FALSE);
}

?>
